src/toil/leader.py

```diff
@@ -516,70 +548,101 @@ def _processReadyJob(self, readyJob, resultStatus):
                 self.issueJob(readyJob)
                 logger.debug("Job: %s is empty, we are scheduling to clean it up", readyJob.jobStoreID)
             else:
-                self.processTotallyFailedJob(readyJob)
+                self.processTotallyFailedJob(job_id)
                 logger.warning("Job: %s is empty but completely failed - something is very wrong", readyJob.jobStoreID)
 
     def _processReadyJobs(self):
         """Process jobs that are ready to be scheduled/have successors to schedule"""
         logger.debug('Built the jobs list, currently have %i jobs to update and %i jobs issued',
-                     len(self.toilState.updatedJobs), self.getNumberOfJobsIssued())
-
-        updatedJobs = self.toilState.updatedJobs # The updated jobs to consider below
-        self.toilState.updatedJobs = {} # Resetting the collection for the next group of updated jobs
-
-        for updatedJob, resultStatus in updatedJobs.values():
-            self._processReadyJob(updatedJob, resultStatus)
+                     self.toilState.bus.count(JobUpdatedMessage), self.getNumberOfJobsIssued())
+
+        # Now go through and, for each job that has updated this tick, process it.
+
+        # We are in trouble if the same job is updated with multiple statuses,
+        # and we want to coalesce multiple updates with the same status in the
+        # same tick, so we remember what we've done so far in this dict from
+        # job ID to status.
+        handled_with_status = {}
+        for message in self.toilState.bus.for_each(JobUpdatedMessage):
+            # Handle all the job update messages that came in.
+
+            if message.job_id in handled_with_status:
+                if handled_with_status[message.job_id] == message.result_status:
+                    # This is a harmless duplicate
+                    logger.debug("Coalescing duplicate update %s", message)
```

Comment:
This might be noisy, even for debugging.  If it actually is useful, it should probably include context about this being a job queue message, or something to tell the user "what" is being updated.

Reply:
We shouldn't be getting loads and loads of these duplicate update messages. But I can make them more specific.